### ISSUE TITLE
Fatal Exception: java.lang.UnsatisfiedLinkError: dlopen failed: "/data/data/com.asda.android/lib-main/libgnustl_shared.so" is 32-bit instead of 64-bit

### ISSUE THREAD

**Issue by vkalmath** [facebook/react-native#18599]
- [x] I have reviewed the [documentation](https://facebook.github.io/react-native)
- [x] I have searched [existing issues](https://github.com/facebook/react-native/issues)
- [x] I am using the [latest React Native version](https://github.com/facebook/react-native/releases)

We are using ReactNative in our app. App is crashing with Fatal Exception: java.lang.UnsatisfiedLinkError: dlopen failed: "/data/data/com.asda.android/lib-main/libgnustl_shared.so" is 32-bit instead of 64-bit only on Sony-Xperia device(Xperia XA/Android 7). Its not happening on other devices
we have removed the 64 bit versions of api from apk by adding following things in build.gradle

```
android {
    compileSdkVersion 27
    buildToolsVersion '27.0.2'

    flavorDimensions "default"

    …….

   defaultConfig {  
        ...  
        ndk {  
            abiFilters "armeabi-v7a", "x86"  
        } 
    }

    packagingOptions {
    ……
        exclude '/lib/mips64/**'
        exclude '/lib/arm64-v8a/**'
        exclude '/lib/x86_64/**'
    }

    …….
}
```

## Environment
Scanning folders for symlinks in /Users/vn0zpn2/ern_workspace/asda-ern-miniapps/DMTMiniApp/node_modules (5ms)

Environment:
  OS: macOS Sierra 10.12.6
  Node: 9.3.0
  Yarn: 1.3.2
  npm: 5.6.0
  Watchman: 4.9.0
  Xcode: Xcode 9.2 Build version 9C40b
  Android Studio: Not Found

Packages: (wanted => installed)
  react: 16.0.0 => 16.0.0
  react-native: 0.51.0 => 0.51.0

Its happening in production. We are using the Litho along with React native(version: 0.51.0)
build.gradle

## Steps to Reproduce
App Launch.

## Expected Behavior
Should not crash on app launch

## Actual Behavior
Crashes with log below
```
Fatal Exception: java.lang.UnsatisfiedLinkError: dlopen failed: "/data/data/com.asda.android/lib-main/libgnustl_shared.so" is 32-bit instead of 64-bit
       at java.lang.Runtime.load0(Runtime.java:897)
       at java.lang.System.load(System.java:1505)
       at com.facebook.soloader.DirectorySoSource.loadLibraryFrom(DirectorySoSource.java:71)
       at com.facebook.soloader.DirectorySoSource.loadLibrary(DirectorySoSource.java:42)
       at com.facebook.soloader.SoLoader.loadLibraryBySoName(SoLoader.java:299)
       at com.facebook.soloader.DirectorySoSource.loadLibraryFrom(DirectorySoSource.java:65)
       at com.facebook.soloader.DirectorySoSource.loadLibrary(DirectorySoSource.java:42)
       at com.facebook.soloader.SoLoader.loadLibraryBySoName(SoLoader.java:299)
       at com.facebook.soloader.DirectorySoSource.loadLibraryFrom(DirectorySoSource.java:65)
       at com.facebook.soloader.DirectorySoSource.loadLibrary(DirectorySoSource.java:42)
       at com.facebook.soloader.SoLoader.loadLibraryBySoName(SoLoader.java:299)
       at com.facebook.soloader.DirectorySoSource.loadLibraryFrom(DirectorySoSource.java:65)
       at com.facebook.soloader.DirectorySoSource.loadLibrary(DirectorySoSource.java:42)
       at com.facebook.soloader.SoLoader.loadLibraryBySoName(SoLoader.java:299)
       at com.facebook.soloader.SoLoader.loadLibrary(SoLoader.java:247)
       at com.facebook.react.bridge.ReactBridge.staticInit(ReactBridge.java:20)
       at com.facebook.react.bridge.NativeMap.(NativeMap.java:21)
       at com.facebook.react.bridge.JSCJavaScriptExecutorFactory.create(JSCJavaScriptExecutorFactory.java:23)
       at com.facebook.react.ReactInstanceManager$4.run(ReactInstanceManager.java:928)
       at java.lang.Thread.run(Thread.java:761)
```

**Comment by vkalmath** [comment]
updated..

**Comment by shijilsmb** [comment]
I have the same problem with Samsung s7 and s8

**Comment by n1ru4l** [comment]
@mindhivefi Could you please provide an example?

**Comment by hramos** [comment]
It looks like this issue has gotten quite stale. With 64-bit now available in master, I am going to close this and request anybody running into something similar to please open a new issue.

**Comment by chandranshu12** [comment]
Wow, ran into this just now and can't believe this was closed just 17 hours ago with no resolution.

**Comment by hramos** [comment]
I am sorry you ran into this issue with such poor timing. Please do open a new issue as advised.

**Comment by n1ru4l** [comment]
Since I  forgot to post a solution for the problem:

One of my dependencies shipped their own `libgnustl_shared.so` which has overwritten the one provided by react-native. I resolved this by moving `implementation "com.facebook.react:react-native:+"` as high up as possible and tewaking the `packagingOptions`

**HOWEVER** I did not upgrade to the newest version with 64bit support yet so you might want to ditch the `exclude everything 64bit related` section.

**`app/build.gradle`**
```
...
android {
    ...
    packagingOptions {
        // exclude everything 64bit related
        exclude '/lib/mips64/**'
        exclude '/lib/arm64-v8a/**'
        exclude '/lib/x86_64/**'

        // com.thirdparty:spotify-player:24 ships their own .so files
        // we do not want those
        pickFirst '/lib/x86/libgnustl_shared.so'
        pickFirst '/lib/armeabi-v7a/libgnustl_shared.so'
    }
}

dependencies {
    ...
    implementation "com.facebook.react:react-native:+"
    ...
    implementation 'com.thirdparty:spotify-player:24'
}
``` 

Maybe this ca help someone. You should start by analysing your dependencies (e.g. if this causes the issue you are facing), before applying this fix.

**Comment by russelRajitha** [comment]
`defaultConfig {
        ndk.abiFilters 'armeabi-v7a','arm64-v8a','x86','x86_64'
        packagingOptions {
            exclude '/lib/mips64/**'
            exclude '/lib/arm64-v8a/**'
     }
}`


this worked for me

**Comment by MriduDixit15** [comment]
> 
> 
> `defaultConfig { ndk.abiFilters 'armeabi-v7a','arm64-v8a','x86','x86_64' packagingOptions { exclude '/lib/mips64/**' exclude '/lib/arm64-v8a/**' } }`
> 
> this worked for me

Did it worked in play store?

**Comment by russelRajitha** [comment]
yes it worked.

**Comment by wangchongwei** [comment]
> yes it worked.

just once, then failed again. sad!

**Comment by russelRajitha** [comment]
> > yes it worked.
> 
> just once, then failed again. sad!

what is the error?

**Comment by JensonMiralles** [comment]
> > > yes it worked.
> > 
> > 
> > just once, then failed again. sad!
> 
> what is the error?

Doesn't work anymore on Playstore :'(

**Comment by russelRajitha** [comment]
> > > > yes it worked.
> > > 
> > > 
> > > just once, then failed again. sad!
> > 
> > 
> > what is the error?
> 
> Doesn't work anymore on Playstore :'(

what is exact issue? is it uploading apk or running?

**Comment by JensonMiralles** [comment]
> > > > > yes it worked.
> > > > 
> > > > 
> > > > just once, then failed again. sad!
> > > 
> > > 
> > > what is the error?
> > 
> > 
> > Doesn't work anymore on Playstore :'(
> 
> what is exact issue? is it uploading apk or running?

It's running alright, but Playstore won't accept it because of the 64 bit issue.

**Comment by russelRajitha** [comment]
i got the same error when running and uploading to playstore. and above code helped me. and my app now on live.

**Comment by webexpert4rv** [comment]
What is your RN version ??? @russelRajitha

**Comment by russelRajitha** [comment]
0.55.2

**Comment by Balajipss** [comment]
Using RN 0.60.5 version. 
If you guys are wondering why your app crashes if you download it from playstore, make the below changes

packagingOptions {
    exclude '/lib/mips64/**'
    exclude '/lib/arm64-v8a/**'
    exclude '/lib/x86_64/**'
    pickFirst '/lib/x86/libc++_shared.so' //In pickFirst, replace the library name with the library that is causing error.
    pickFirst '/lib/armeabi-v7a/libc++_shared.so' In pickFirst, replace the library name with the library that is causing error.
}

REASON - We are excluding all 64 bit libraries so that OS will not have confusion of which library to pick.

If you do not know which library is causing the crash, upload your bundle in Firebase test lab and then check. It will tell you which library is throwing the error

**Comment by ppv94** [comment]
I am using react-native version 56.0 I am able to build apk but when I try to upload in play store it doesn't work. what should i do

here is my build.gradlew

`    compileSdkVersion 27
    buildToolsVersion '27.0.3'
    defaultConfig {
        applicationId "com.christenpro"
        minSdkVersion 16
        targetSdkVersion 27
        versionCode 32
        versionName "2.24"

        vectorDrawables.useSupportLibrary = true
        ndk {
            abiFilters "armeabi-v7a", "x86"
        }
        multiDexEnabled = true
       dexOptions {
           jumboMode true
            javaMaxHeapSize "4g"
       // preDexLibraries = false
    }
configurations.all {
    resolutionStrategy.eachDependency { DependencyResolveDetails details ->
        if (details.getRequested().getGroup() == 'com.google.android.gms') {
            details.useVersion('15.0.1')
        }
    }
}
    }`

**Comment by JensonMiralles** [comment]
> I am using react-native version 56.0 I am able to build apk but when I try to upload in play store it doesn't work. what should i do
> 
> here is my build.gradlew
> 
> ` compileSdkVersion 27
> buildToolsVersion '27.0.3'
> defaultConfig {
> applicationId "com.christenpro"
> minSdkVersion 16
> targetSdkVersion 27
> versionCode 32
> versionName "2.24"
> 
> ```
>     vectorDrawables.useSupportLibrary = true
>     ndk {
>         abiFilters "armeabi-v7a", "x86"
>     }
>     multiDexEnabled = true
>    dexOptions {
>        jumboMode true
>         javaMaxHeapSize "4g"
>    // preDexLibraries = false
> }
> ```
> 
> configurations.all {
> resolutionStrategy.eachDependency { DependencyResolveDetails details ->
> if (details.getRequested().getGroup() == 'com.google.android.gms') {
> details.useVersion('15.0.1')
> }
> }
> }
> }`

Our ultimate fix was to upgrade to 0.59 react-native version
